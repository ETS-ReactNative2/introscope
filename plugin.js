const toPairs = obj => {
    const pairs = []
    for (const key in obj) {
        pairs.push([key, obj[key]])
    }
    return pairs
}

const byType = type => node => node.type == type
const not = fn => (...args) => !fn(...args)

function processProgram({ types: t }, programPath, programState) {
    let options = { removeImports: false }

    const scopeId = programPath.scope.generateUidIdentifier('scope')

    const scopeSet = (left, right) =>
        t.assignmentExpression('=', t.memberExpression(scopeId, left), right)

    const unwrapOrRemoveExports = path => {
        path.get('body').forEach(path => {
            if (path.isExportDefaultDeclaration()) {
                path.replaceWith(
                    t.expressionStatement(
                        scopeSet(
                            t.identifier('default'),
                            path.get('declaration').node
                        )
                    )
                )
            } else if (path.isExportNamedDeclaration()) {
                path.node.declaration
                    ? path.replaceWith(path.get('declaration').node)
                    : path.remove()
            } else if (path.isExportAllDeclaration()) {
                path.remove()
            }
        })
    }

    const identifiersToObjectProperties = identifiers =>
        identifiers.map(identifier =>
            t.objectProperty(identifier, identifier, undefined, true)
        )

    const wrapInFunction = (globalIds, body) =>
        t.functionExpression(
            null,
            [t.assignmentPattern(scopeId, t.objectExpression([]))],
            t.blockStatement(
                []
                    .concat(
                        globalIds.length
                            ? t.variableDeclaration(
                                  'var',
                                  globalIds.map(id => t.variableDeclarator(id))
                              )
                            : []
                    )
                    .concat(body)
                    .concat([t.returnStatement(scopeId)])
            )
        )

    const getGlobalIdentifiers = scope =>
        toPairs(scope.globals).map(([_, identifier]) => identifier)

    const moduleExports = right =>
        t.expressionStatement(
            t.assignmentExpression(
                '=',
                t.memberExpression(
                    t.identifier('module'),
                    t.identifier('exports')
                ),
                right
            )
        )

    const variableDeclaratorToScope = (path, identifier) => {
        const idPath = path.get('id')
        if (idPath.isObjectPattern()) {
            path.insertAfter(
                t.variableDeclarator(
                    path.scope.generateUidIdentifier('temp'),
                    scopeSet(t.clone(identifier), t.clone(identifier))
                )
            )
        } else if (idPath.isArrayPattern()) {
        } else {
            const init = path.get('init')
            if (!init.node) return
            // Q: Why not replace the parent VariableDeclaration node?
            // A: To not die debugging variable declaration order in a `for` loop if one of the binding should be ignored dew to introscope configuration (expected future feature). And it's simplier to implement :)
            init.replaceWith(scopeSet(path.get('id').node, init.node))
        }
    }

    const classDeclarationToScope = path => {
        const classExpression = t.clone(path.node)
        classExpression.type = 'ClassExpression'
        path.replaceWith(
            t.expressionStatement(
                scopeSet(path.get('id').node, classExpression)
            )
        )
    }

    const functionDeclarationToScope = path => {
        const program = path.findParent(path => path.isProgram())
        program.unshiftContainer(
            'body',
            t.expressionStatement(
                scopeSet(path.get('id').node, path.get('id').node)
            )
        )
    }

    const declarationToScope = (path, identifier) => {
        if (path.isNodeType('VariableDeclarator')) {
            variableDeclaratorToScope(path, identifier)
        } else if (path.isNodeType('ClassDeclaration')) {
            classDeclarationToScope(path)
        } else if (path.isNodeType('FunctionDeclaration')) {
            functionDeclarationToScope(path)
        } else if (
            path.isNodeType('ImportDefaultSpecifier') ||
            path.isNodeType('ImportSpecifier') ||
            path.isNodeType('ImportNamespaceSpecifier')
        ) {
            if (options.removeImports === true) {
                // ignore
            } else if (path.parentPath.isImportDeclaration()) {
                return path.get('local').node
            }
        } else {
            throw new TypeError('Unknown node.type = ' + path.node.type)
            // TODO: log warning here using babel logger
        }
    }

    const replaceReferenceWithScope = path => {
        // ExportNamedDeclaration gets properly processed by replaceMutationWithScope()
        if (path.node.type == 'ExportNamedDeclaration') return
        // ExportSpecifier gets removed by unwrapOrRemoveExports()
        if (path.parent.type == 'ExportSpecifier') return
        if (path.findParent(path => path.isObjectPattern())) return

        const scoped = t.memberExpression(scopeId, path.node)
        if (path.parent && path.parent.type == 'CallExpression') {
            path.replaceWith(
                t.sequenceExpression([t.numericLiteral(0), scoped])
            )
        } else {
            path.replaceWith(scoped)
        }
    }

    const replaceMutationWithScope = path => {
        if (path.node.type == 'AssignmentExpression') {
            const left = path.get('left')
            left.replaceWith(t.memberExpression(scopeId, left.node))
        }
    }

    const bindingToScope = binding => {
        binding.referencePaths.forEach(replaceReferenceWithScope)
        binding.constantViolations.forEach(replaceMutationWithScope)
        return declarationToScope(binding.path, binding.identifier)
    }

    const bindingsToScope = bindings =>
        toPairs(bindings)
            .map(([_, binding]) => binding)
            .map(bindingToScope)
            .filter(Boolean)

    const program = (path, state) => {
        const globalIds = getGlobalIdentifiers(path.scope)
        const programGlobalNames = Object.keys(path.scope.globals)

        const localImportIds = bindingsToScope(path.scope.bindings)

        // unwrapOrRemoveExports() should go after bindingsToScope() as the latter treats `export var/let/const` as a reference and uses node.parent to distinguish
        unwrapOrRemoveExports(path)

        // reverse()-ing to preserve order after unshift()-ing
        localImportIds
            .reverse()
            .forEach(localId =>
                path.node.body.unshift(
                    t.expressionStatement(scopeSet(localId, localId))
                )
            )

        const programBody = path.node.body
        if (options.removeImports === true) {
            path.node.body = []
        } else {
            const importsOnly = programBody.filter(byType('ImportDeclaration'))
            path.node.body = importsOnly
        }
        const bodyWithoutImports = programBody.filter(
            not(byType('ImportDeclaration'))
        )

        path.pushContainer(
            'body',
            moduleExports(wrapInFunction(globalIds, bodyWithoutImports))
        )

        const uniqueGlobalIds = new Set()
        path.traverse({
            Scope(path, state) {
                programGlobalNames.forEach(globalName => {
                    const binding = path.scope.getBinding(globalName)
                    uniqueGlobalIds.add(binding)
                })
            }
        })
        bindingsToScope(Array.from(uniqueGlobalIds.values()))
    }

    program(programPath, programState)
}

export default function(babel) {
    return {
        visitor: {
            Program(path, state) {
                processProgram(babel, path, state)
            }
        }
    }
}
