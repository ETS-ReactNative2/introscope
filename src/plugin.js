const toPairs = obj => {
    const pairs = [];
    for (const key in obj) {
        pairs.push([key, obj[key]]);
    }
    return pairs;
};

const byType = type => node => node.type == type;
const not = fn => (...args) => !fn(...args);

function processProgram({ types: t }, programPath, programState) {
    const defaultOptions = {
        enable: false,
        ignore: [],
        removeImports: false
    };

    let options = {
        ...defaultOptions,
        ...programState.opts
    };

    const scopeId = programPath.scope.generateUidIdentifier('scope');

    const scopeSet = (left, right) =>
        t.assignmentExpression('=', t.memberExpression(scopeId, left), right);

    const unwrapOrRemoveExports = path => {
        path.get('body').forEach(path => {
            if (path.isExportDefaultDeclaration()) {
                path.replaceWith(
                    t.expressionStatement(
                        scopeSet(
                            t.identifier('default'),
                            path.get('declaration').node
                        )
                    )
                );
            } else if (path.isExportNamedDeclaration()) {
                path.node.declaration
                    ? path.replaceWith(path.get('declaration').node)
                    : path.remove();
            } else if (path.isExportAllDeclaration()) {
                path.remove();
            }
        });
    };

    const identifiersToObjectProperties = identifiers =>
        identifiers.map(identifier =>
            t.objectProperty(identifier, identifier, undefined, true)
        );

    const wrapInFunction = (globalIds, body) =>
        t.functionExpression(
            null,
            [t.assignmentPattern(scopeId, t.objectExpression([]))],
            t.blockStatement(
                []
                    .concat(
                        globalIds.length
                            ? t.variableDeclaration(
                                  'var',
                                  globalIds.map(id => t.variableDeclarator(id))
                              )
                            : []
                    )
                    .concat(body)
                    .concat([t.returnStatement(scopeId)])
            )
        );

    const moduleExports = right =>
        t.expressionStatement(
            t.assignmentExpression(
                '=',
                t.memberExpression(
                    t.identifier('module'),
                    t.identifier('exports')
                ),
                right
            )
        );

    const variableDeclaratorToScope = (path, identifier) => {
        const idPath = path.get('id');
        if (idPath.isObjectPattern() || idPath.isArrayPattern()) {
            path.insertAfter(
                t.variableDeclarator(
                    path.scope.generateUidIdentifier('temp'),
                    scopeSet(t.clone(identifier), t.clone(identifier))
                )
            );
        } else {
            const init = path.get('init');
            if (!init.node) return;
            // Q: Why not replace the parent VariableDeclaration node?
            // A: To not die debugging variable declaration order in a `for` loop if one of the binding should be ignored dew to introscope configuration (expected future feature). And it's simplier to implement :)
            init.replaceWith(scopeSet(path.get('id').node, init.node));
        }
    };

    const classDeclarationToScope = path => {
        const classExpression = t.clone(path.node);
        classExpression.type = 'ClassExpression';
        path.replaceWith(
            t.expressionStatement(
                scopeSet(path.get('id').node, classExpression)
            )
        );
    };

    const functionDeclarationToScope = path => {
        const program = path.findParent(path => path.isProgram());
        program.unshiftContainer(
            'body',
            t.expressionStatement(
                scopeSet(path.get('id').node, path.get('id').node)
            )
        );
    };

    const declarationToScope = (path, identifier) => {
        if (path.isNodeType('VariableDeclarator')) {
            variableDeclaratorToScope(path, identifier);
        } else if (path.isNodeType('ClassDeclaration')) {
            classDeclarationToScope(path);
        } else if (path.isNodeType('FunctionDeclaration')) {
            functionDeclarationToScope(path);
        } else if (
            path.isNodeType('ImportDefaultSpecifier') ||
            path.isNodeType('ImportSpecifier') ||
            path.isNodeType('ImportNamespaceSpecifier')
        ) {
            if (options.removeImports === true) {
                // ignore
            } else if (path.parentPath.isImportDeclaration()) {
                return path.get('local').node;
            }
        } else {
            throw path.buildCodeFrameError(
                'Cannot apply declarationToScope() to an anknown node.type: ' +
                    path.node.type
            );
        }
    };

    const replaceReferenceWithScope = path => {
        // ExportNamedDeclaration gets properly processed by replaceMutationWithScope()
        if (path.node.type == 'ExportNamedDeclaration') return;
        // ExportSpecifier gets removed by unwrapOrRemoveExports()
        if (path.parent.type == 'ExportSpecifier') return;
        if (path.findParent(path => path.isObjectPattern())) return;

        const scoped = t.memberExpression(scopeId, path.node);
        if (path.parent && path.parent.type == 'CallExpression') {
            path.replaceWith(
                t.sequenceExpression([t.numericLiteral(0), scoped])
            );
        } else {
            path.replaceWith(scoped);
        }
    };

    const replaceMutationWithScope = path => {
        if (path.node.type == 'AssignmentExpression') {
            const left = path.get('left');
            left.replaceWith(t.memberExpression(scopeId, left.node));
        }
    };

    const bindingToScope = binding => {
        binding.referencePaths.forEach(replaceReferenceWithScope);
        binding.constantViolations.forEach(replaceMutationWithScope);
        return declarationToScope(binding.path, binding.identifier);
    };

    const bindingsToScope = bindings =>
        toPairs(bindings)
            .filter(([name, _]) => !options.ignore.includes(name))
            .map(([_, binding]) => binding)
            .map(bindingToScope)
            .filter(Boolean);

    const parseConfig = path => {
        const firstStatement = path.get('body').forEach(statement => {
            const leading = statement.node.leadingComments || [];
            const trailing = statement.node.trailingComments || [];
            const comments = [].concat(leading).concat(trailing);
            comments.map(node => node.value).forEach(comment => {
                const [_, configJSON] = comment.split(/^\s*@introscope-config/);
                if (configJSON) {
                    try {
                        const config = JSON.parse(`{${configJSON}}`);
                        options = {
                            ...options,
                            ...config
                        };
                    } catch (ex) {
                        console.error(
                            'Error parsing Introscope config:',
                            comment
                        );
                    }
                }
            });
        });
    };

    const program = (path, state) => {
        parseConfig(path);
        if (!options.enable) return;

        const globalIds = toPairs(path.scope.globals)
            .filter(([name, _]) => !options.ignore.includes(name))
            .map(([_, identifier]) => identifier);
        const programGlobalNames = Object.keys(path.scope.globals);

        const localImportIds = bindingsToScope(path.scope.bindings);

        // unwrapOrRemoveExports() should go after bindingsToScope() as the latter treats `export var/let/const` as a reference and uses node.parent to distinguish
        unwrapOrRemoveExports(path);

        // reverse()-ing to preserve order after unshift()-ing
        localImportIds
            .reverse()
            .forEach(localId =>
                path.node.body.unshift(
                    t.expressionStatement(scopeSet(localId, localId))
                )
            );

        const programBody = path.node.body;
        if (options.removeImports === true) {
            path.node.body = [];
        } else {
            const importsOnly = programBody.filter(byType('ImportDeclaration'));
            path.node.body = importsOnly;
        }
        const bodyWithoutImports = programBody.filter(
            not(byType('ImportDeclaration'))
        );

        path.pushContainer(
            'body',
            moduleExports(wrapInFunction(globalIds, bodyWithoutImports))
        );

        // globals become locals after wrapInFunction()
        // TODO: make it more explicit in the code
        const globalBindings = {};
        path.traverse({
            Scope(path, state) {
                programGlobalNames.forEach(globalName => {
                    const binding = path.scope.getBinding(globalName);
                    globalBindings[globalName] = binding;
                });
            }
        });
        bindingsToScope(globalBindings);
    };

    program(programPath, programState);
}

module.exports = function(babel) {
    return {
        visitor: {
            Program(path, state) {
                if (state.opts.disable) return;

                if (
                    typeof process == 'object' &&
                    process.env.NODE_ENV != 'test'
                ) {
                    return;
                }

                processProgram(babel, path, state);
            }
        }
    };
};
