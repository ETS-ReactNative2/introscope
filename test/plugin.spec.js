import { transform } from 'babel-core';
import plugin from '../babel-plugin';

const shoot = (code, opts = {}) =>
    expect(
        transform(code, {
            sourceType: 'module',
            plugins: [
                'syntax-flow',
                'syntax-object-rest-spread',
                [plugin, { enable: true, ...opts }],
            ],
        }).code,
    ).toMatchSnapshot();

describe('plugin', () => {
    it('import', () => {
        const code = `
            import 'some-module1'
            import defaultImport from 'some-module2'
            import { singleNamedImport } from 'some-module3'
            import { namedImport1, namedImport2 } from 'some-module4'
            import { originalImportName as localImportName } from 'some-module5'
            import * as namespaceImport from 'some-module6'

            [
                defaultImport,
                singleNamedImport,
                namedImport1,
                namedImport2,
                localImportName,
                namespaceImport
            ]
        `;
        shoot(code);
        shoot(code, { removeImports: true });
    });

    it('assignments', () => {
        shoot(`
            let x;
            x++;
            x = 1;
        `);
    });

    it('references', () => {
        shoot(`
            let x;
            x();
            !function(){ x++ };
            () => { return { x } };
            !{ x }
        `);
    });

    it('declarations', () => {
        shoot(`
            let noInitValue;
            for (var forVar = 0; forVar != 0; forVar++);
            let singleInit = 1;
            let doubleInit1 = 1, doubleInit2 = 2;
            var sameDoubleInit = 1, sameDoubleInit = 2;

            let { singleDestructing } = 1
            var { objectDestruction1, objectDestruction2, nestedProperty: { objectNestedDestruction1 } } = { objectDestruction1, objectDestruction2, nestedProperty: { objectNestedDestruction1 } }, let2 = objectDestruction1;
            let { ...objectRest } = 1;
            var [ arrayDestruction1, arrayDestruction2, [ arrayNestedDestruction ] ] = [ arrayDestruction1, arrayDestruction2, [ arrayNestedDestruction ] ];
            let [ ...arrayRest ] = 1;


            class ClassName {}; new ClassName();
            functionName(); function functionName () {}
        `);
    });

    it('scopes', () => {
        shoot(`
            let x = true;
            for (let x = 0; x != 0; x++);
            { let x = false; }
            !function(){ var x = false; }
        `);
    });

    it('globals', () => {
        shoot(`
            let x = global1;
            global2 = 1;
            global2 = 1;
            globalFunction();
            !function(){ return [nestedGlobal1, global2] }();
        `);
    });

    it('export', () => {
        shoot(`
            let defaultExport, toBeNameExport1, toBeNameExport2;
            export default defaultExport;
            export const namedSingleExport = 'namedSingleExportValue'
            export { toBeNameExport1, toBeNameExport2 }
            export * from 'some-module'
        `);
    });

    it('flow', () => {
        shoot(`
            import type { TypeImportedTypeShouldBeIgnored } from 'x'
            import { type ImportedTypeShouldBeIgnored } from 'y'
            type LocalTypeShouldBeIgnored = ImportedTypeShouldBeIgnored | TypeImportedTypeShouldBeIgnored;
        `);

        shoot(`
            type SomeType = number;
            type SomeOtherType = SomeType;
            function typedFuntion(x: SomeType): SomeOtherType {
                let typedVar: SomeType | SomeOtherType = 123
            }
        `);

        shoot(`
            opaque type OpaqueType = string;
            let foo: OpaqueType = "foo"
        `);
    });
});

describe('options', () => {
    it('does nothing if NODE_ENV != "test"', () => {
        const NODE_ENV = process.env.NODE_ENV;
        process.env.NODE_ENV = 'production';
        shoot(`
            var shouldBeUntouched = true;
        `);
        process.env.NODE_ENV = 'development';
        shoot(`
        var shouldBeUntouched = true;
        `);
        process.env.NODE_ENV = 'test';
        shoot(`
            var shouldBeTransformed = true;
        `);
        process.env.NODE_ENV = NODE_ENV;
    });

    it('removeImports', () => {
        shoot(`
            // @introscope "removeImports": true
            // other comment
            import sholdBeRemoved from 'some-module1'
            sholdBeRemoved++
        `);
        shoot(`
            // @other-at-comment bla bla bla
            import sholdBeRemoved from 'some-module1'
            sholdBeRemoved++
            // @introscope "removeImports": true
        `);
        shoot(
            `
            // @introscope "removeImports": false
            import sholdNotBeRemoved from 'some-module1'
            sholdNotBeRemoved++
        `,
            { removeImports: true },
        );
        shoot(`
            // @introscope "removeImports": true
            import sholdNotBeRemoved from 'some-module1'
            sholdNotBeRemoved++
            // last takes precedence
            // @introscope "removeImports": false
        `);
        shoot(`
            import sholdNotBeRemoved from 'some-module1'
            defaultImport1++
        `);
    });

    describe('ignore', () => {
        it('adds', () => {
            shoot(`
            // @introscope "ignore": ["localIgnored", "deepGlobalIgnored"]
            const x = { deep: { global: { variable: deepGlobal, variableIgnored: deepGlobalIgnored }}}
            const localIgnored = 1;
            localIgnored++;
            function throwError (message) {
                throw message
            }
            `);
        });

        it('removes', () => {
            shoot(`
            // @introscope "ignore": ["-Error", "-Date"]
            [
                Error, Date
            ]
            `);
        });

        it('id', () => {
            shoot(`
            // @introscope "ignore": []
            [
                Error, Date
            ]
            `);
        });

        it('ignores builtins', () => {
            shoot(`
            [
                Infinity,
                NaN,
                undefined,
                null,
                eval,
                isFinite,
                isNaN,
                parseFloat,
                parseInt,
                decodeURI,
                decodeURIComponent,
                encodeURI,
                encodeURIComponent,
                escape,
                unescape,
                Object,
                Function,
                Boolean,
                Symbol,
                Error,
                EvalError,
                InternalError,
                RangeError,
                ReferenceError,
                SyntaxError,
                TypeError,
                URIError,
                Number,
                Math,
                Date,
                String,
                RegExp,
                Array,
                Int8Array,
                Uint8Array,
                Uint8ClampedArray,
                Int16Array,
                Uint16Array,
                Int32Array,
                Uint32Array,
                Float32Array,
                Float64Array,
                Map,
                Set,
                WeakMap,
                WeakSet,
                SIMD,
                ArrayBuffer,
                SharedArrayBuffer, 
                Atomics, 
                DataView,
                JSON,
                Promise,
                Generator,
                GeneratorFunction,
                AsyncFunction,
                Reflect,
                Proxy,
                Intl,
                WebAssembly,
                arguments,
                Buffer,
            ]
            `);
        });
    });

    it('enable', () => {
        shoot(
            `
            let shouldBeUntouched = true;
        `,
            { enable: false },
        );

        shoot(
            `
            // @introscope "enable": false
            let shouldBeUntouched = true;
        `,
            { enable: true },
        );

        shoot(
            `
            // @introscope "enable": true
            let shouldBeTransformed = true;
        `,
            { enable: false },
        );
    });

    it('disable', () => {
        shoot(
            `
            let shouldBeUntouched = true;
        `,
            { disable: true },
        );

        shoot(
            `
            // inline "disable" is ignored, use "enable": false
            // @introscope "disable": true
            let shouldBeTransformed = true;
        `,
        );

        shoot(
            `
            let shouldBeTransformed = true;
        `,
        );
    });

    describe('exportName', () => {
        it('uses default', () => {
            shoot(`
                // no exportName
                `);
        });

        it('works with magic comment', () => {
            shoot(
                `
                // @introscope "exportName": "microscope1"
                `,
            );
        });

        it('works with plugin options', () => {
            shoot(
                `
                // code
                `,
                { exportName: 'microscope' },
            );
        });
    });
});

describe('introscope import', () => {
    it('should stay', () => {
        shoot(
            `
            import { foo } from 'foo';
            import { scope } from 'introscope';
            export const introscope = scope({
                constantA,
                functionB,
                // other identifiers of your module
            });
        `,
            { removeImports: true },
        );
    });
});

describe('test', () => {
    const code = `
    import { introscope } from 'directModule';
    import { introscope as renamed } from 'renamedModule';
    import { otherNamedImport } from 'otheModule';
    `;

    it('does normal transpilation if enabled in opts', () => {
        shoot(code, { enable: true });
    });

    it('does normal transpilation if enabled in code', () => {
        shoot(code + '\n// @introscope "enable": true', {
            enable: false,
        });
    });

    it('in test code does transpile only introscope import', () => {
        shoot(code, { enable: false });
    });

    it('in test code does transpile only introscope import', () => {
        shoot(code, { enable: false, instrumentImports: false });
    });
});
