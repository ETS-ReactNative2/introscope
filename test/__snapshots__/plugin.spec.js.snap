// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options disable 1`] = `
"
let shouldBeUntouched = true;"
`;

exports[`options disable 2`] = `
"export const introscope = function (_scope = {}) {
            // inline \\"disable\\" is ignored, use \\"enable\\": false
            // @introscope \\"disable\\": true
            let shouldBeTransformed = _scope.shouldBeTransformed = true;
            return _scope;
};"
`;

exports[`options disable 3`] = `
"export const introscope = function (_scope = {}) {
            let shouldBeTransformed = _scope.shouldBeTransformed = true;
            return _scope;
};"
`;

exports[`options does nothing if NODE_ENV != "test" 1`] = `
"
var shouldBeUntouched = true;"
`;

exports[`options does nothing if NODE_ENV != "test" 2`] = `
"
var shouldBeUntouched = true;"
`;

exports[`options does nothing if NODE_ENV != "test" 3`] = `
"export const introscope = function (_scope = {}) {
            var shouldBeTransformed = _scope.shouldBeTransformed = true;
            return _scope;
};"
`;

exports[`options enable 1`] = `
"
let shouldBeUntouched = true;"
`;

exports[`options enable 2`] = `
"
// @introscope \\"enable\\": false
let shouldBeUntouched = true;"
`;

exports[`options enable 3`] = `
"export const introscope = function (_scope = {}) {
            // @introscope \\"enable\\": true
            let shouldBeTransformed = _scope.shouldBeTransformed = true;
            return _scope;
};"
`;

exports[`options exportName uses default 1`] = `
"
// no exportName
export const introscope = function (_scope = {}) {
                return _scope;
};"
`;

exports[`options exportName works with magic comment 1`] = `
"
// @introscope \\"exportName\\": \\"microscope1\\"
export const microscope1 = function (_scope = {}) {
                return _scope;
};"
`;

exports[`options exportName works with plugin options 1`] = `
"
// code
export const microscope = function (_scope = {}) {
                return _scope;
};"
`;

exports[`options ignore 1`] = `
"export const introscope = function (_scope = {}) {
    var deepGlobal;
    _scope.throwError = throwError;

    // @introscope \\"ignore\\": [\\"localIgnored\\", \\"Error\\", \\"deepGlobalIgnored\\"]
    const x = _scope.x = { deep: { global: { variable: _scope.deepGlobal, variableIgnored: deepGlobalIgnored } } };
    const localIgnored = 1;
    localIgnored++;
    function throwError(message) {
        throw new Error(message);
    }
    return _scope;
};"
`;

exports[`options removeImports 1`] = `
"export const introscope = function (_scope = {}) {
            _scope.sholdBeRemoved++;
            return _scope;
};"
`;

exports[`options removeImports 2`] = `
"export const introscope = function (_scope = {}) {
            _scope.sholdBeRemoved++;
            // @introscope \\"removeImports\\": true

            return _scope;
};"
`;

exports[`options removeImports 3`] = `
"
// @introscope \\"removeImports\\": false
import sholdNotBeRemoved from 'some-module1';
export const introscope = function (_scope = {}) {
            _scope.sholdNotBeRemoved = sholdNotBeRemoved;
            _scope.sholdNotBeRemoved++;
            return _scope;
};"
`;

exports[`options removeImports 4`] = `
"
// @introscope \\"removeImports\\": true
import sholdNotBeRemoved from 'some-module1';
export const introscope = function (_scope = {}) {
            _scope.sholdNotBeRemoved = sholdNotBeRemoved;
            _scope.sholdNotBeRemoved++;
            // last takes precedence
            // @introscope \\"removeImports\\": false

            return _scope;
};"
`;

exports[`options removeImports 5`] = `
"
import sholdNotBeRemoved from 'some-module1';
export const introscope = function (_scope = {}) {
            var defaultImport1;
            _scope.sholdNotBeRemoved = sholdNotBeRemoved;
            _scope.defaultImport1++;
            return _scope;
};"
`;

exports[`plugin assignments 1`] = `
"export const introscope = function (_scope = {}) {
            let x;
            _scope.x++;
            _scope.x = 1;
            return _scope;
};"
`;

exports[`plugin declarations 1`] = `
"export const introscope = function (_scope = {}) {
            _scope.functionName = functionName;

            let noInitValue;
            for (var forVar = _scope.forVar = 0; _scope.forVar != 0; _scope.forVar++);
            let singleInit = _scope.singleInit = 1;
            let doubleInit1 = _scope.doubleInit1 = 1,
                doubleInit2 = _scope.doubleInit2 = 2;
            var sameDoubleInit = 1,
                sameDoubleInit = _scope.sameDoubleInit = 2;

            let { singleDestructing } = 1,
                _temp = _scope.singleDestructing = singleDestructing;
            var { objectDestruction1, objectDestruction2, nestedProperty: { objectNestedDestruction1 } } = { objectDestruction1: _scope.objectDestruction1, objectDestruction2: _scope.objectDestruction2, nestedProperty: { objectNestedDestruction1: _scope.objectNestedDestruction1 } },
                _temp4 = _scope.objectNestedDestruction1 = objectNestedDestruction1,
                _temp3 = _scope.objectDestruction2 = objectDestruction2,
                _temp2 = _scope.objectDestruction1 = objectDestruction1,
                let2 = _scope.let2 = _scope.objectDestruction1;
            let { ...objectRest } = 1,
                _temp5 = _scope.objectRest = objectRest;
            var [arrayDestruction1, arrayDestruction2, [arrayNestedDestruction]] = [_scope.arrayDestruction1, _scope.arrayDestruction2, [_scope.arrayNestedDestruction]],
                _temp8 = _scope.arrayNestedDestruction = arrayNestedDestruction,
                _temp7 = _scope.arrayDestruction2 = arrayDestruction2,
                _temp6 = _scope.arrayDestruction1 = arrayDestruction1;
            let [...arrayRest] = 1,
                _temp9 = _scope.arrayRest = arrayRest;

            _scope.ClassName = class ClassName {};
            ;new _scope.ClassName();
            (0, _scope.functionName)();function functionName() {}
            return _scope;
};"
`;

exports[`plugin export 1`] = `
"export const introscope = function (_scope = {}) {
            let defaultExport, toBeNameExport1, toBeNameExport2;
            _scope.default = _scope.defaultExport;
            const namedSingleExport = _scope.namedSingleExport = 'namedSingleExportValue';
            return _scope;
};"
`;

exports[`plugin flow 1`] = `
"
import type { TypeImportedTypeShouldBeIgnored } from 'x';
import { type ImportedTypeShouldBeIgnored } from 'y';
export const introscope = function (_scope = {}) {
            type LocalTypeShouldBeIgnored = ImportedTypeShouldBeIgnored | TypeImportedTypeShouldBeIgnored;
            return _scope;
};"
`;

exports[`plugin flow 2`] = `
"export const introscope = function (_scope = {}) {
    _scope.typedFuntion = typedFuntion;

    type SomeType = number;
    type SomeOtherType = SomeType;
    function typedFuntion(x: SomeType): SomeOtherType {
        let typedVar: SomeType | SomeOtherType = 123;
    }
    return _scope;
};"
`;

exports[`plugin globals 1`] = `
"export const introscope = function (_scope = {}) {
            var global2, global1, globalFunction, nestedGlobal1;

            let x = _scope.x = _scope.global1;
            _scope.global2 = 1;
            _scope.global2 = 1;
            (0, _scope.globalFunction)();
            !function () {
                        return [_scope.nestedGlobal1, _scope.global2];
            }();
            return _scope;
};"
`;

exports[`plugin import 1`] = `
"
import 'some-module1';
import defaultImport from 'some-module2';
import { singleNamedImport } from 'some-module3';
import { namedImport1, namedImport2 } from 'some-module4';
import { originalImportName as localImportName } from 'some-module5';
import * as namespaceImport from 'some-module6';

export const introscope = function (_scope = {}) {
    _scope.defaultImport = defaultImport;
    _scope.singleNamedImport = singleNamedImport;
    _scope.namedImport1 = namedImport1;
    _scope.namedImport2 = namedImport2;
    _scope.localImportName = localImportName;
    _scope.namespaceImport = namespaceImport;
    [_scope.defaultImport, _scope.singleNamedImport, _scope.namedImport1, _scope.namedImport2, _scope.localImportName, _scope.namespaceImport];
    return _scope;
};"
`;

exports[`plugin import 2`] = `
"export const introscope = function (_scope = {}) {

    [_scope.defaultImport, _scope.singleNamedImport, _scope.namedImport1, _scope.namedImport2, _scope.localImportName, _scope.namespaceImport];
    return _scope;
};"
`;

exports[`plugin references 1`] = `
"export const introscope = function (_scope = {}) {
            let x;
            (0, _scope.x)();
            !function () {
                        _scope.x++;
            };
            () => {
                        return { x: _scope.x };
            };
            !{ x: _scope.x };
            return _scope;
};"
`;

exports[`plugin scopes 1`] = `
"export const introscope = function (_scope = {}) {
            let x = _scope.x = true;
            for (let x = 0; x != 0; x++);
            {
                        let x = false;
            }
            !function () {
                        var x = false;
            };
            return _scope;
};"
`;

exports[`test does normal transpilation if enabled in code 1`] = `
"
import { introscope } from 'directModule';
import { introscope as renamed } from 'renamedModule';
import { otherNamedImport } from 'otheModule';

// @introscope \\"enable\\": true

export const introscope = function (_scope = {}) {
    _scope.introscope = introscope;
    _scope.renamed = renamed;
    _scope.otherNamedImport = otherNamedImport;
    return _scope;
};"
`;

exports[`test does normal transpilation if enabled in opts 1`] = `
"
import { introscope } from 'directModule';
import { introscope as renamed } from 'renamedModule';
import { otherNamedImport } from 'otheModule';
export const introscope = function (_scope = {}) {
    _scope.introscope = introscope;
    _scope.renamed = renamed;
    _scope.otherNamedImport = otherNamedImport;
    return _scope;
};"
`;

exports[`test in test code does transpile only introscope import 1`] = `
"
import { introscope } from 'directModule?introscope';
import { introscope as renamed } from 'renamedModule?introscope';
import { otherNamedImport } from 'otheModule';"
`;

exports[`test in test code does transpile only introscope import 2`] = `
"
import { introscope } from 'directModule';
import { introscope as renamed } from 'renamedModule';
import { otherNamedImport } from 'otheModule';"
`;
